src/Weft/Server.hs

```diff
@@ -19,50 +20,78 @@ import Network.HTTP.Types.Method
 import Data.Aeson hiding (json)
 import qualified Data.ByteString.Char8 as C8
 import Data.Monoid
+import Data.Maybe
 import qualified Data.ByteString.Lazy as BL
 import Control.Monad.Reader
 import Data.Text (Text)
+import qualified Data.Text as T
+import qualified Data.HashMap.Strict as HM
+import qualified Data.Map as M
 
 
 parseReqBody :: (Wefty query)
-             => Text
+             => ClientRequest
              -> Either String ((Gql query m s) 'Query)
-parseReqBody
+parseReqBody (ClientRequest q v _)
   = first errorBundlePretty
-  . parse (runReaderT queryParser mempty) "<server>"
+  . parse (runReaderT queryParser v) "<server>" $ q
+
+data ClientRequest = 
+  ClientRequest { queryContent  :: Text
+                , variables     :: Vars
+                , operationName :: Maybe Text
+                } deriving (Show, Eq, Ord)
+
+maybeClientRequest :: C8.ByteString -> Maybe ClientRequest
+maybeClientRequest rb = do
+  json  <- decode @Value $ BL.fromStrict rb
+  q     <- json ^? key "query" . _String
+  let v = fromMaybe M.empty $ do
+            vars     <- json ^? key "variables" . _Object
+            pure     $  M.fromList . stringify . HM.toList $ vars
+  let o = json ^? key "operationName" . _String
+  pure $ ClientRequest q v o
+
+stringify :: [(Text, Value)] -> [(String, String)]
+stringify = fmap f
+  where
+    f (t,v) = ( T.unpack t
+              , C8.unpack . BL.toStrict . encode $ v
+              )
 
-maybeQuery :: C8.ByteString -> Maybe Text
-maybeQuery rb = do
-  json <- decode @Value $ BL.fromStrict rb
-  json ^? key "query" . _String
 
 note :: Maybe a -> Either String a
-note Nothing = Left ""
+note Nothing = Left "maybeClientRequest failed"
```

Comment:
make `note :: String -> Maybe a -> Either String a`, then you don't have to bake this string in here